experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,130 @@
+where
+    F: Send + Sync + Clone + 'static + FnOnce(Request) -> O,
+    O: Send + Sync + Future<Output = Response>,
+{
+    pub fn new(anonymity_value: usize, request_handler: F) -> Self {
+        Self {
+            anonymity_value,
+            current_requests: Arc::new(Mutex::new(vec![])),
+            request_handler,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (response_sender, response_receiver) = oneshot::channel();
+
+        let request = Message {
+            data: request,
+            response_sender,
+        };
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut current_requests = self
+                .current_requests
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+            current_requests.push(request);
+
+            if current_requests.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(current_requests.deref_mut(), vec![]);
+
+                // Shuffle current requests and send them to the backend.
+                let request_handler = self.request_handler.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_handler).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        response_receiver.await.context("Couldn't receive response")
+    }
+
+    // Lexicographically sorts requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(
+        mut requests: Vec<Message>,
+        request_handler: F,
+    ) -> anyhow::Result<()> {
+        requests.sort_by(|first, second| first.data.cmp(&second.data));
+
+        let response_futures = requests.into_iter().map(|request| {
+            let request_handler_clone = request_handler.clone();
+            async move {
+                let response = (request_handler_clone)(request.data).await;
+                Message {
+                    data: response,
+                    response_sender: request.response_sender,
+                }
+            }
+        });
+        let responses = join_all(response_futures).await;
+
+        Self::shuffle_responses(responses)
+    }
+
+    // Lexicographically sorts received backend responses and sends them back to the client async
+    // tasks using corresponding channels defined in the [`Message::response_sender`].
+    fn shuffle_responses(mut responses: Vec<Message>) -> anyhow::Result<()> {
+        responses.sort_by(|first, second| first.data.cmp(&second.data));
```

Comment:
But if we don't shuffle back, wouldn't the network attacker also deasonymize clients: because the responses will arrive to Oak and depart from Oak with the same order?